protocols/imap/src/main/java/org/apache/james/imap/processor/base/UidMsnMapper.java

```diff
@@ -75,7 +79,41 @@ public synchronized void clear() {
     }
 
     public synchronized void addUid(MessageUid uid) {
-        this.addMapping(nextMsn(), uid);
+        if (isLastUid(uid)) {
+            addMapping(nextMsn(), uid);
+        } else {
+            addUidInMiddle(uid);
+        }
+    }
+
+    private boolean isLastUid(MessageUid uid) {
+        return msnToUid.isEmpty() ||
+            uid.asLong() > msnToUid.get(getLastMsn()).asLong();
+    }
+
+    private boolean alreadyContains(MessageUid uid) {
+        return msnToUid.containsValue(uid);
+    }
+
+    private void addUidInMiddle(MessageUid uid) {
+        List<MessageUid> aboveUids = removeAndGetAboveUidSortedInIncreasingOrder(uid);
+        addMapping(nextMsn(), uid);
+        for (MessageUid aboveUid : aboveUids) {
+            addMapping(nextMsn(), aboveUid);
+        }
+    }
+
+    private List<MessageUid> removeAndGetAboveUidSortedInIncreasingOrder(MessageUid uid) {
+        ImmutableList.Builder<MessageUid> result = ImmutableList.builder();
+        int position = getLastMsn();
+        Optional<MessageUid> maxUid = getUid(position);
+        while (maxUid.isPresent() && uid.asLong() < maxUid.get().asLong()) {
```

Comment:
The previous code was using not only one, but two TreeMap.

The previous code is still vulnerable to this issue, as key for entries is generated with a **monotic** counter, it's still vulnerable to out of order events.

TreeMap IMO don't bring any benefit:
 - Two structure to maintain
 - The fact that it's a *sorted* structure (red-back tree) don't bring any benefit, as:
      - It's sorted on the keys generated from a monotic counter
      - It don't offer me which key have a special value (What BiMap does)
      - It don't handle the key generation

I'll try to have a look for simpler solution.

As a general remark, for such kind of thing, we tend to propose commits in the team, to avoid very long discussions...